File: web/lib/xp-progress.ts
import {
  XP_ACTIONS,
  createFreshXpState,
  type XpAction,
  type XpAwardRecord,
  type XpState,
} from "@/lib/xp";

export type XpOpportunity = {
  code: string;
  label: string;
  xp: number;
  href: string;
  priority: number;
};

export type XpWeekProgress = {
  planId: string;
  weekId: string;
  weekIndex: number | null;
  phaseLabel: string;
  startDate: string;
  endDate: string;
  completedSessions: number;
  plannedSessions: number;
  remainingSessions: number;
  complete: boolean;
  weekXpEarned: boolean;
};

export type XpMilestone = {
  id: string;
  planId: string;
  milestoneType: "phase_completed" | "plan_completed" | "camp_completed";
  phaseLabel: string | null;
  completedAt: string;
  displayLabel: string;
};

export type XpProgress = {
  state: XpState;
  opportunities: XpOpportunity[];
  currentWeek: XpWeekProgress | null;
  majorMilestones: XpMilestone[];
};

const actionNames = new Set(Object.keys(XP_ACTIONS));
const milestoneTypes = new Set(["phase_completed", "plan_completed", "camp_completed"]);

function record(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

function nonNegativeInteger(value: unknown): number | null {
  return typeof value === "number" && Number.isSafeInteger(value) && value >= 0
    ? value
    : null;
}

function validDateTime(value: unknown): value is string {
  return typeof value === "string" && value.trim().length > 0 && Number.isFinite(Date.parse(value));
}

function parseAward(value: unknown): XpAwardRecord | null {
  const candidate = record(value);
  if (!candidate) return null;
  const action = candidate.action;
  const amount = nonNegativeInteger(candidate.amount);
  if (
    typeof candidate.id !== "string" ||
    !candidate.id.trim() ||
    typeof action !== "string" ||
    !actionNames.has(action) ||
    amount === null ||
    amount !== XP_ACTIONS[action as XpAction].xp ||
    !validDateTime(candidate.awarded_at)
  ) {
    return null;
  }
  const calendarDate =
    typeof candidate.calendar_date === "string" && /^\d{4}-\d{2}-\d{2}$/.test(candidate.calendar_date)
      ? candidate.calendar_date
      : undefined;
  return {
    id: candidate.id.trim(),
    action: action as XpAction,
    amount,
    awardedAt: new Date(candidate.awarded_at).toISOString(),
    ...(calendarDate ? { calendarDate } : {}),
  };
}

function parseState(value: unknown): XpState {
  const candidate = record(value);
  if (!candidate) throw new Error("Server returned invalid XP progress.");
  const totalXp = nonNegativeInteger(candidate.total_xp);
  const awards = Array.isArray(candidate.recent_awards)
    ? candidate.recent_awards.map(parseAward)
    : [];
  if (
    totalXp === null ||
    !Array.isArray(candidate.recent_awards) ||
    awards.some((award) => award === null) ||
    awards.length > 20
  ) {
    throw new Error("Server returned invalid XP progress.");
  }
  const lastDailyLoginDate =
    candidate.last_daily_login_date === null || candidate.last_daily_login_date === undefined
      ? null
      : typeof candidate.last_daily_login_date === "string" &&
          /^\d{4}-\d{2}-\d{2}$/.test(candidate.last_daily_login_date)
        ? candidate.last_daily_login_date
        : undefined;
  if (lastDailyLoginDate === undefined) {
    throw new Error("Server returned invalid XP progress.");
  }
  return {
    ...createFreshXpState(),
    totalXp,
    lastDailyLoginDate,
    recentAwards: awards as XpAwardRecord[],
  };
}

function parseOpportunity(value: unknown): XpOpportunity | null {
  const candidate = record(value);
  if (!candidate) return null;
  const xp = nonNegativeInteger(candidate.xp);
  const priority = nonNegativeInteger(candidate.priority);
  if (
    typeof candidate.code !== "string" ||
    !candidate.code.trim() ||
    typeof candidate.label !== "string" ||
    !candidate.label.trim() ||
    typeof candidate.href !== "string" ||
    !candidate.href.startsWith("/") ||
    xp === null ||
    xp <= 0 ||
    priority === null
  ) {
    return null;
  }
  return {
    code: candidate.code.trim(),
    label: candidate.label.trim(),
    href: candidate.href,
    xp,
    priority,
  };
}

function parseWeek(value: unknown): XpWeekProgress | null {
  if (value === null || value === undefined) return null;
  const candidate = record(value);
  if (!candidate) throw new Error("Server returned invalid XP week progress.");
  const completedSessions = nonNegativeInteger(candidate.completed_sessions);
  const plannedSessions = nonNegativeInteger(candidate.planned_sessions);
  const remainingSessions = nonNegativeInteger(candidate.remaining_sessions);
  const rawWeekIndex = candidate.week_index;
  const weekIndex = rawWeekIndex === null || rawWeekIndex === undefined
    ? null
    : nonNegativeInteger(rawWeekIndex);
  if (
    typeof candidate.plan_id !== "string" ||
    !candidate.plan_id.trim() ||
    typeof candidate.week_id !== "string" ||
    !candidate.week_id.trim() ||
    typeof candidate.phase_label !== "string" ||
    typeof candidate.start_date !== "string" ||
    typeof candidate.end_date !== "string" ||
    completedSessions === null ||
    plannedSessions === null ||
    remainingSessions === null ||
    (rawWeekIndex !== null && rawWeekIndex !== undefined && weekIndex === null) ||
    typeof candidate.complete !== "boolean" ||
    typeof candidate.week_xp_earned !== "boolean" ||
    completedSessions > plannedSessions ||
    remainingSessions !== plannedSessions - completedSessions
  ) {
    throw new Error("Server returned invalid XP week progress.");
  }
  return {
    planId: candidate.plan_id.trim(),
    weekId: candidate.week_id.trim(),
    weekIndex,
    phaseLabel: candidate.phase_label.trim(),
    startDate: candidate.start_date,
    endDate: candidate.end_date,
    completedSessions,
    plannedSessions,
    remainingSessions,
    complete: candidate.complete,
    weekXpEarned: candidate.week_xp_earned,
  };
}

function parseMilestone(value: unknown): XpMilestone | null {
  const candidate = record(value);
  if (!candidate) return null;
  const milestoneType = candidate.milestone_type;
  if (
    typeof candidate.id !== "string" ||
    typeof candidate.plan_id !== "string" ||
    typeof milestoneType !== "string" ||
    !milestoneTypes.has(milestoneType) ||
    !validDateTime(candidate.completed_at) ||
    typeof candidate.display_label !== "string" ||
    !candidate.display_label.trim() ||
    !(
      candidate.phase_label === null ||
      candidate.phase_label === undefined ||
      typeof candidate.phase_label === "string"
    )
  ) {
    return null;
  }
  return {
    id: candidate.id,
    planId: candidate.plan_id,
    milestoneType: milestoneType as XpMilestone["milestoneType"],
    phaseLabel: typeof candidate.phase_label === "string" ? candidate.phase_label : null,
    completedAt: new Date(candidate.completed_at).toISOString(),
    displayLabel: candidate.display_label.trim(),
  };
}

export function createFreshXpProgress(): XpProgress {
  return {
    state: createFreshXpState(),
    opportunities: [],
    currentWeek: null,
    majorMilestones: [],
  };
}

export function parseXpProgressResponse(value: unknown): XpProgress {
  const candidate = record(value);
  if (!candidate) throw new Error("Server returned invalid XP progress.");
  const opportunities = Array.isArray(candidate.opportunities)
    ? candidate.opportunities.map(parseOpportunity)
    : [];
  const milestones = Array.isArray(candidate.major_milestones)
    ? candidate.major_milestones.map(parseMilestone)
    : [];
  if (
    !Array.isArray(candidate.opportunities) ||
    opportunities.some((item) => item === null) ||
    opportunities.length > 3 ||
    !Array.isArray(candidate.major_milestones) ||
    milestones.some((item) => item === null) ||
    milestones.length > 50
  ) {
    throw new Error("Server returned invalid XP progress.");
  }
  return {
    state: parseState(candidate.state),
    opportunities: opportunities as XpOpportunity[],
    currentWeek: parseWeek(candidate.current_week),
    majorMilestones: milestones as XpMilestone[],
  };
}

File: api/services/xp_progress.py
"""Read-only athlete XP progress, opportunities and milestone presentation.

The interface must never invent earning opportunities in the browser. This
module derives them from persisted profile/intake/plan/check-in/completion state
and the same Today/week contracts that own the rest of the product.
"""

from __future__ import annotations

import logging
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Any, Literal

from api.services.today_readiness_boundary import build_today_command_view
from api.services.today_service import resolve_training_day
from api.services.week_progress import evaluate_week_completion, find_week_for_training_day
from api.services.streaks import get_streak_state
from api.services.xp_awards import profile_activation_complete
from api.store import AppStore

logger = logging.getLogger(__name__)

RECENT_AWARDS_LIMIT = 20
MILESTONES_LIMIT = 50
COMPLETED_SESSION_STATUSES = frozenset({"done", "modified"})
RecordReadStatus = Literal["found", "not_found", "unavailable"]


@dataclass(frozen=True)
class _RecordRead:
    """Result of an optional-record read without collapsing failure into absence."""

    status: RecordReadStatus
    value: Mapping[str, Any] | None = None


def _rows(value: object) -> list[dict[str, Any]]:
    if not isinstance(value, Sequence) or isinstance(value, (str, bytes, bytearray)):
        return []
    return [dict(row) for row in value if isinstance(row, Mapping)]


def _response_data(response: object) -> object:
    return getattr(response, "data", None)


def _client(store: AppStore) -> object | None:
    return getattr(store, "client", None)


def _read_xp_state(store: AppStore, athlete_id: str) -> dict[str, Any]:
    custom = getattr(store, "get_xp_progress_state", None)
    if callable(custom):
        result = custom(athlete_id, limit=RECENT_AWARDS_LIMIT)
        if isinstance(result, Mapping):
            return {
                "total_xp": max(0, int(result.get("total_xp") or 0)),
                "last_daily_login_date": result.get("last_daily_login_date"),
                "recent_awards": _rows(result.get("recent_awards")),
            }

    # Test/in-memory stores already keep the durable account and ledger shapes.
    accounts = getattr(store, "xp_accounts", None)
    awards_by_athlete = getattr(store, "xp_awards", None)
    if isinstance(accounts, Mapping) and isinstance(awards_by_athlete, Mapping):
        account = accounts.get(athlete_id)
        account_map = account if isinstance(account, Mapping) else {}
        awards = _rows(awards_by_athlete.get(athlete_id))
        awards.sort(
            key=lambda row: (
                str(row.get("awarded_at") or ""),
                str(row.get("id") or ""),
            ),
            reverse=True,
        )
        return {
            "total_xp": max(0, int(account_map.get("total_xp") or 0)),
            "last_daily_login_date": account_map.get("last_daily_login_date"),
            "recent_awards": [
                {
                    key: row[key]
                    for key in (
                        "id",
                        "action",
                        "amount",
                        "awarded_at",
                        "calendar_date",
                    )
                    if row.get(key) is not None
                }
                for row in awards[:RECENT_AWARDS_LIMIT]
            ],
        }

    client = _client(store)
    if client is None:
        raise RuntimeError("XP progress store is unavailable")

    account_response = (
        client.table("xp_accounts")
        .select("total_xp,last_daily_login_date")
        .eq("athlete_id", athlete_id)
        .limit(1)
        .execute()
    )
    account_rows = _rows(_response_data(account_response))
    account = account_rows[0] if account_rows else {}

    awards_response = (
        client.table("xp_awards")
        .select("id,action,amount,awarded_at,calendar_date")
        .eq("athlete_id", athlete_id)
        .order("awarded_at", desc=True)
        .order("id", desc=True)
        .limit(RECENT_AWARDS_LIMIT)
        .execute()
    )
    return {
        "total_xp": max(0, int(account.get("total_xp") or 0)),
        "last_daily_login_date": account.get("last_daily_login_date"),
        "recent_awards": _rows(_response_data(awards_response)),
    }


def _list_milestones(store: AppStore, athlete_id: str) -> list[dict[str, Any]]:
    custom = getattr(store, "list_plan_milestones", None)
    if callable(custom):
        return _rows(custom(athlete_id, limit=MILESTONES_LIMIT))

    in_memory = getattr(store, "plan_milestones", None)
    if isinstance(in_memory, Mapping):
        values = in_memory.get(athlete_id)
        milestones = _rows(values)
    elif isinstance(in_memory, Sequence) and not isinstance(
        in_memory,
        (str, bytes, bytearray),
    ):
        milestones = [
            dict(row)
            for row in in_memory
            if isinstance(row, Mapping)
            and str(row.get("athlete_id") or "") == athlete_id
        ]
    else:
        milestones = []

    if milestones:
        milestones.sort(
            key=lambda row: str(row.get("completed_at") or ""),
            reverse=True,
        )
        return milestones[:MILESTONES_LIMIT]

    client = _client(store)
    if client is None:
        return []
    response = (
        client.table("plan_milestones")
        .select(
            "id,plan_id,milestone_type,milestone_key,phase_label,metadata,completed_at"
        )
        .eq("athlete_id", athlete_id)
        .order("completed_at", desc=True)
        .limit(MILESTONES_LIMIT)
        .execute()
    )
    return _rows(_response_data(response))


def _award_exists(
    store: AppStore,
    athlete_id: str,
    *,
    action: str | None = None,
    idempotency_key: str | None = None,
    calendar_date: str | None = None,
) -> bool:
    awards_by_athlete = getattr(store, "xp_awards", None)
    if isinstance(awards_by_athlete, Mapping):
        for row in _rows(awards_by_athlete.get(athlete_id)):
            if action is not None and str(row.get("action") or "") != action:
                continue
            if (
                idempotency_key is not None
                and str(row.get("idempotency_key") or "") != idempotency_key
            ):
                continue
            if (
                calendar_date is not None
                and str(row.get("calendar_date") or "") != calendar_date
            ):
                continue
            return True
        return False

    client = _client(store)
    if client is None:
        return False
    query = client.table("xp_awards").select("id").eq("athlete_id", athlete_id)
    if action is not None:
        query = query.eq("action", action)
    if idempotency_key is not None:
        query = query.eq("idempotency_key", idempotency_key)
    if calendar_date is not None:
        query = query.eq("calendar_date", calendar_date)
    return bool(_rows(_response_data(query.limit(1).execute())))


def _optional_record_read(row: object, *, source: str, athlete_id: str) -> _RecordRead:
    if row is None:
        return _RecordRead(status="not_found")
    if isinstance(row, Mapping):
        return _RecordRead(status="found", value=dict(row))
    logger.error(
        "[xp] malformed %s progress read athlete_id=%s type=%s",
        source,
        athlete_id,
        type(row).__name__,
    )
    return _RecordRead(status="unavailable")


def _safe_latest_intake(store: AppStore, athlete_id: str) -> _RecordRead:
    try:
        return _optional_record_read(
            store.get_latest_intake(athlete_id),
            source="intake",
            athlete_id=athlete_id,
        )
    except Exception:  # noqa: BLE001 - activation opportunities fail closed
        logger.exception("[xp] intake progress read failed athlete_id=%s", athlete_id)
        return _RecordRead(status="unavailable")


def _safe_latest_plan(store: AppStore, athlete_id: str) -> _RecordRead:
    try:
        return _optional_record_read(
            store.get_latest_plan(athlete_id),
            source="plan",
            athlete_id=athlete_id,
        )
    except Exception:  # noqa: BLE001 - activation opportunities fail closed
        logger.exception("[xp] plan progress read failed athlete_id=%s", athlete_id)
        return _RecordRead(status="unavailable")


def _today_command(
    store: AppStore,
    *,
    athlete_id: str,
    athlete_timezone: str | None,
) -> object | None:
    try:
        return build_today_command_view(
            store,
            athlete_id=athlete_id,
            athlete_timezone=athlete_timezone,
        )
    except Exception:  # noqa: BLE001 - opportunities fail closed
        logger.exception("[xp] Today progress read failed athlete_id=%s", athlete_id)
        return None


def _active_plan(
    store: AppStore,
    *,
    athlete_id: str,
    command: object | None,
) -> Mapping[str, Any] | None:
    command_plan = getattr(command, "active_plan", None)
    plan_id = (
        str(command_plan.get("id") or "").strip()
        if isinstance(command_plan, Mapping)
        else ""
    )
    if not plan_id:
        return None
    try:
        row = store.get_plan_for_athlete(plan_id, athlete_id)
        return row if isinstance(row, Mapping) else None
    except Exception:  # noqa: BLE001 - progress fails closed
        logger.exception(
            "[xp] active plan progress read failed athlete_id=%s plan_id=%s",
            athlete_id,
            plan_id,
        )
        return None


def _current_week(
    store: AppStore,
    *,
    athlete_id: str,
    plan: Mapping[str, Any] | None,
    training_day: str,
) -> dict[str, Any] | None:
    if not plan:
        return None
    week = find_week_for_training_day(plan, training_day)
    if not isinstance(week, Mapping):
        return None
    plan_id = str(plan.get("id") or "").strip()
    week_id = str(week.get("week_id") or "").strip()
    start_date = str(week.get("start_date") or "").strip()
    end_date = str(week.get("end_date") or "").strip()
    if not plan_id or not week_id or not start_date or not end_date:
        return None
    try:
        completions = store.list_plan_session_completions(
            athlete_id,
            plan_id,
            limit=500,
        )
    except Exception:  # noqa: BLE001 - progress fails closed
        logger.exception(
            "[xp] week completion read failed athlete_id=%s plan_id=%s",
            athlete_id,
            plan_id,
        )
        return None
    relevant = [
        row
        for row in _rows(completions)
        if start_date <= str(row.get("training_day") or "") <= end_date
    ]
    evaluation = evaluate_week_completion(week=week, completions=relevant)
    planned = max(0, int(evaluation.get("planned") or 0))
    skipped = len(evaluation.get("skipped_session_ids") or [])
    resolved = max(0, int(evaluation.get("resolved") or 0))
    completed = max(0, min(planned, resolved - skipped))
    week_xp_earned = _award_exists(
        store,
        athlete_id,
        action="full_training_week_completed",
        calendar_date=start_date,
    )
    return {
        "plan_id": plan_id,
        "week_id": week_id,
        "week_index": week.get("week_index"),
        "phase_label": str(week.get("phase_label") or "").strip(),
        "start_date": start_date,
        "end_date": end_date,
        "completed_sessions": completed,
        "planned_sessions": planned,
        "remaining_sessions": max(0, planned - completed),
        "complete": bool(evaluation.get("complete")),
        "week_xp_earned": week_xp_earned,
    }


def _milestone_view(row: Mapping[str, Any]) -> dict[str, Any] | None:
    milestone_type = str(row.get("milestone_type") or "").strip()
    if milestone_type == "phase_completed":
        phase_label = str(row.get("phase_label") or "Training").strip()
        display_label = f"{phase_label} phase complete"
    elif milestone_type == "plan_completed":
        display_label = "Plan complete"
    elif milestone_type == "camp_completed":
        display_label = "Fight camp complete"
    else:
        return None
    return {
        "id": str(row.get("id") or "").strip(),
        "plan_id": str(row.get("plan_id") or "").strip(),
        "milestone_type": milestone_type,
        "phase_label": row.get("phase_label"),
        "completed_at": row.get("completed_at"),
        "display_label": display_label,
    }


def _opportunity(
    *,
    code: str,
    label: str,
    xp: int,
    href: str,
    priority: int,
) -> dict[str, Any]:
    return {
        "code": code,
        "label": label,
        "xp": xp,
        "href": href,
        "priority": priority,
    }


def _opportunities(
    store: AppStore,
    *,
    athlete_id: str,
    profile: object,
    latest_intake: _RecordRead,
    latest_plan: _RecordRead,
    command: object | None,
    current_week: Mapping[str, Any] | None,
) -> list[dict[str, Any]]:
    choices: list[dict[str, Any]] = []

    if (
        not _award_exists(store, athlete_id, action="profile_completed")
        and not profile_activation_complete(profile)
    ):
        choices.append(
            _opportunity(
                code="complete_profile",
                label="Complete your athlete profile",
                xp=25,
                href="/settings",
                priority=50,
            )
        )
    if (
        latest_intake.status == "not_found"
        and not _award_exists(store, athlete_id, action="first_intake_completed")
    ):
        choices.append(
            _opportunity(
                code="complete_first_intake",
                label="Complete your first intake",
                xp=50,
                href="/onboarding",
                priority=60,
            )
        )
    if (
        latest_plan.status == "not_found"
        and not _award_exists(store, athlete_id, action="first_plan_ready")
    ):
        choices.append(
            _opportunity(
                code="build_first_plan",
                label="Build your first training plan",
                xp=100,
                href="/onboarding",
                priority=70,
            )
        )

    today = getattr(command, "today", None)
    training_day = str(getattr(today, "training_day", "") or "").strip()
    recommendation_state = str(
        getattr(today, "recommendation_state", "") or ""
    ).strip()
    if recommendation_state == "not_checked_in":
        first_checkin_earned = _award_exists(
            store,
            athlete_id,
            action="first_checkin_completed",
        )
        choices.append(
            _opportunity(
                code="complete_today_checkin",
                label="Complete today's check-in",
                xp=10 if first_checkin_earned else 35,
                href="/today",
                priority=10,
            )
        )

    open_injuries = getattr(command, "open_injuries", None)
    if (
        training_day
        and isinstance(open_injuries, Sequence)
        and any(isinstance(injury, Mapping) for injury in open_injuries)
        and not _award_exists(
            store,
            athlete_id,
            action="injury_update_completed",
            calendar_date=training_day,
        )
    ):
        choices.append(
            _opportunity(
                code="update_active_injury",
                label="Update today's active injury",
                xp=10,
                href="/today",
                priority=20,
            )
        )

    session_scope = str(getattr(today, "session_scope", "") or "")
    next_session = getattr(today, "next_session", None)
    completion_status = str(getattr(today, "completion_status", "") or "")
    decision_tier = str(getattr(today, "decision_tier", "") or "")
    if (
        session_scope == "today"
        and isinstance(next_session, Mapping)
        and bool(next_session)
        and completion_status not in COMPLETED_SESSION_STATUSES
        and decision_tier != "stop"
    ):
        choices.append(
            _opportunity(
                code="complete_today_session",
                label=(
                    "Complete today's modified session"
                    if decision_tier == "modify"
                    else "Complete today's session"
                ),
                xp=75,
                href="/today",
                priority=30,
            )
        )

    if (
        current_week
        and int(current_week.get("planned_sessions") or 0) > 0
        and not bool(current_week.get("complete"))
        and not bool(current_week.get("week_xp_earned"))
    ):
        choices.append(
            _opportunity(
                code="complete_training_week",
                label="Complete this training week",
                xp=100,
                href="/progress",
                priority=40,
            )
        )

    choices.sort(key=lambda item: (int(item["priority"]), str(item["code"])))
    return choices[:3]


def build_xp_progress(
    store: AppStore,
    *,
    athlete_id: str,
    athlete_timezone: str | None,
    profile: object,
) -> dict[str, Any]:
    """Build XP progress and record the authenticated app day idempotently."""

    state = _read_xp_state(store, athlete_id)
    latest_intake = _safe_latest_intake(store, athlete_id)
    latest_plan = _safe_latest_plan(store, athlete_id)
    command = _today_command(
        store,
        athlete_id=athlete_id,
        athlete_timezone=athlete_timezone,
    )
    training_day = str(
        getattr(getattr(command, "today", None), "training_day", "")
        or resolve_training_day(athlete_timezone)
    )
    plan = _active_plan(store, athlete_id=athlete_id, command=command)
    current_week = _current_week(
        store,
        athlete_id=athlete_id,
        plan=plan,
        training_day=training_day,
    )
    milestones = [
        view
        for row in _list_milestones(store, athlete_id)
        if (view := _milestone_view(row)) is not None
    ]
    opportunities = _opportunities(
        store,
        athlete_id=athlete_id,
        profile=profile,
        latest_intake=latest_intake,
        latest_plan=latest_plan,
        command=command,
        current_week=current_week,
    )
    return {
        "state": state,
        "streaks": get_streak_state(
            store,
            athlete_id=athlete_id,
            athlete_timezone=athlete_timezone,
        ),
        "opportunities": opportunities,
        "current_week": current_week,
        "major_milestones": milestones,
    }


__all__ = ["build_xp_progress"]
